include/swift/Runtime/Concurrent.h

```diff
@@ -406,6 +407,91 @@ class ConcurrentMap
   }
 };
 
+
+template <class ElemTy> struct ConcurrentReadableArray {
+private:
+  struct Storage {
+    std::atomic<size_t> Count;
+    typename std::aligned_storage<sizeof(ElemTy), alignof(ElemTy)>::type Elem;
+    
+    ElemTy *data() {
+      return reinterpret_cast<ElemTy *>(&Elem);
+    }
+  };
+  
+  std::atomic<size_t> Capacity;
+  std::atomic<size_t> ReaderCount;
+  std::atomic<Storage *> Elements;
+  Mutex WriterLock;
+  std::vector<Storage *> FreeList;
+  
+  Storage *allocate(size_t capacity) {
+    auto size = sizeof(Storage) + (capacity - 1) * sizeof(Storage().Elem);
+    auto *ptr = reinterpret_cast<Storage *>(malloc(size));
+    ptr->Count.store(0, std::memory_order_relaxed);
+    return ptr;
+  }
+  
+  void deallocate(Storage *storage) {
+    if (storage == nullptr) return;
+    
+    auto *data = storage->data();
+    for (size_t i = 0; i < storage->Count; i++) {
+      data[i].~ElemTy();
+    }
+    free(storage);
+  }
+  
+  
+public:
+  void push_back(const ElemTy &elem) {
+    ScopedLock guard(WriterLock);
+    
+    auto *storage = Elements.load(std::memory_order_relaxed);
+    if (storage == nullptr) {
+      storage = allocate(16);
+      Capacity = 16;
```

Comment:
Isn't this a `seqcst` write? (And below too, and the various reads of `Capacity` and `storage->Count`) In fact, I suspect `Capacity` doesn't need to be atomic at all, since it's protected by `WriterLock` like `FreeList`?